File: kernel/crates/gctl-cli/src/commands/board.rs
//! `gctl board` commands — project management and kanban.

use anyhow::Result;
use gctl_storage::DuckDbStore;

fn open_store(db_path: &str) -> Result<DuckDbStore> {
    Ok(DuckDbStore::open(db_path)?)
}

/// Create a new project.
pub fn create_project(name: &str, key: &str, db_path: &str) -> Result<()> {
    let store = open_store(db_path)?;
    let project = gctl_core::BoardProject {
        id: uuid::Uuid::new_v4().to_string(),
        name: name.to_string(),
        key: key.to_uppercase(),
        counter: 0,
    };
    store.create_board_project(&project)?;
    println!("Created project: {} ({})", project.name, project.key);
    Ok(())
}

/// List projects.
pub fn list_projects(format: &str, db_path: &str) -> Result<()> {
    let store = open_store(db_path)?;
    let projects = store.list_board_projects()?;

    if format == "json" {
        println!("{}", serde_json::to_string_pretty(&projects)?);
        return Ok(());
    }

    if projects.is_empty() {
        println!("No projects. Use `gctl board create-project <name> <key>` to get started.");
        return Ok(());
    }

    println!("{:<36} {:<20} {:<6} {:>5}", "ID", "NAME", "KEY", "COUNT");
    println!("{}", "-".repeat(70));
    for p in &projects {
        println!("{:<36} {:<20} {:<6} {:>5}", p.id, p.name, p.key, p.counter);
    }
    Ok(())
}

/// Create a new issue.
pub fn create_issue(
    project_key: &str,
    title: &str,
    description: Option<&str>,
    priority: &str,
    labels: Option<&str>,
    created_by: &str,
    db_path: &str,
) -> Result<()> {
    let store = open_store(db_path)?;

    // Find project by key
    let projects = store.list_board_projects()?;
    let project = projects
        .iter()
        .find(|p| p.key == project_key.to_uppercase())
        .ok_or_else(|| anyhow::anyhow!("project not found: {}", project_key))?;

    let counter = store.increment_project_counter(&project.id)?;
    let now = chrono::Utc::now();
    let labels_vec: Vec<String> = labels
        .map(|l| l.split(',').map(|s| s.trim().to_string()).filter(|s| !s.is_empty()).collect())
        .unwrap_or_default();

    let issue = gctl_core::BoardIssue {
        id: format!("{}-{}", project.key, counter),
        project_id: project.id.clone(),
        title: title.to_string(),
        description: description.map(String::from),
        status: gctl_core::IssueStatus::Backlog,
        priority: priority.to_string(),
        assignee_id: None,
        assignee_name: None,
        assignee_type: None,
        labels: labels_vec,
        parent_id: None,
        created_at: now,
        updated_at: now,
        created_by_id: created_by.to_string(),
        created_by_name: created_by.to_string(),
        created_by_type: "human".to_string(),
        blocked_by: vec![],
        blocking: vec![],
        session_ids: vec![],
        total_cost_usd: 0.0,
        total_tokens: 0,
        pr_numbers: vec![],
    };

    store.insert_board_issue(&issue)?;
    println!("{} — {}", issue.id, issue.title);
    Ok(())
}

/// List issues with optional filters.
pub fn list_issues(
    project: Option<&str>,
    status: Option<&str>,
    assignee: Option<&str>,
    format: &str,
    db_path: &str,
) -> Result<()> {
    let store = open_store(db_path)?;

    // Resolve project key to ID if provided
    let project_id = if let Some(key) = project {
        let projects = store.list_board_projects()?;
        projects.iter()
            .find(|p| p.key == key.to_uppercase())
            .map(|p| p.id.clone())
    } else {
        None
    };

    let filter = gctl_core::BoardIssueFilter {
        project_id,
        status: status.map(String::from),
        assignee_id: assignee.map(String::from),
        ..Default::default()
    };

    let issues = store.list_board_issues(&filter)?;

    if format == "json" {
        println!("{}", serde_json::to_string_pretty(&issues)?);
        return Ok(());
    }

    if issues.is_empty() {
        println!("No issues found.");
        return Ok(());
    }

    println!("{:<12} {:<12} {:<30} {:<10} {:>8}", "ID", "STATUS", "TITLE", "ASSIGNEE", "COST");
    println!("{}", "-".repeat(76));
    for issue in &issues {
        let assignee = issue.assignee_name.as_deref().unwrap_or("-");
        let title = if issue.title.len() > 28 {
            format!("{}…", &issue.title[..27])
        } else {
            issue.title.clone()
        };
        println!(
            "{:<12} {:<12} {:<30} {:<10} ${:>7.2}",
            issue.id, issue.status.as_str(), title, assignee, issue.total_cost_usd,
        );
    }
    println!("\n{} issues", issues.len());
    Ok(())
}

/// Show a single issue with details.
pub fn show_issue(issue_id: &str, db_path: &str) -> Result<()> {
    let store = open_store(db_path)?;
    let issue = store
        .get_board_issue(issue_id)?
        .ok_or_else(|| anyhow::anyhow!("issue not found: {}", issue_id))?;

    println!("{} — {}", issue.id, issue.title);
    println!("Status:   {}", issue.status.as_str());
    println!("Priority: {}", issue.priority);
    if let Some(ref name) = issue.assignee_name {
        println!("Assignee: {} ({})", name, issue.assignee_type.as_deref().unwrap_or("?"));
    }
    if !issue.labels.is_empty() {
        println!("Labels:   {}", issue.labels.join(", "));
    }
    if let Some(ref desc) = issue.description {
        println!("\n{}", desc);
    }
    if !issue.blocked_by.is_empty() {
        println!("\nBlocked by: {}", issue.blocked_by.join(", "));
    }
    if !issue.session_ids.is_empty() {
        println!("\nSessions: {} (${:.2}, {} tokens)", issue.session_ids.len(), issue.total_cost_usd, issue.total_tokens);
    }

    // Show events
    let events = store.list_board_events(&issue.id)?;
    if !events.is_empty() {
        println!("\nEvents:");
        for event in &events {
            println!("  {} — {} by {}", event.event_type, event.timestamp.format("%Y-%m-%d %H:%M"), event.actor_name);
        }
    }

    // Show comments
    let comments = store.list_board_comments(&issue.id)?;
    if !comments.is_empty() {
        println!("\nComments:");
        for comment in &comments {
            println!("  {} ({}): {}", comment.author_name, comment.created_at.format("%Y-%m-%d %H:%M"), comment.body);
        }
    }

    Ok(())
}

/// Move an issue to a new status.
pub fn move_issue(issue_id: &str, status: &str, actor: &str, db_path: &str) -> Result<()> {
    let store = open_store(db_path)?;
    store.update_board_issue_status(issue_id, status, actor, actor, "human")?;
    println!("{} → {}", issue_id, status);
    Ok(())
}

/// Assign an issue.
pub fn assign_issue(issue_id: &str, assignee_name: &str, assignee_type: &str, db_path: &str) -> Result<()> {
    let store = open_store(db_path)?;
    let assignee_id = assignee_name.to_lowercase().replace(' ', "-");
    store.assign_board_issue(issue_id, &assignee_id, assignee_name, assignee_type)?;
    println!("{} → assigned to {} ({})", issue_id, assignee_name, assignee_type);
    Ok(())
}

/// Add a comment to an issue.
pub fn comment(issue_id: &str, body: &str, author: &str, db_path: &str) -> Result<()> {
    let store = open_store(db_path)?;
    let comment = gctl_core::BoardComment {
        id: uuid::Uuid::new_v4().to_string(),
        issue_id: issue_id.to_string(),
        author_id: author.to_string(),
        author_name: author.to_string(),
        author_type: "human".to_string(),
        body: body.to_string(),
        created_at: chrono::Utc::now(),
        session_id: None,
    };
    store.insert_board_comment(&comment)?;
    println!("Comment added to {}", issue_id);
    Ok(())
}

File: shell/gctrl-shell/src/commands/board.ts
import { Command, Options, Args } from "@effect/cli"
import { Console, Effect, Option, Schema } from "effect"
import { KernelClient } from "../services/KernelClient"
import { GhIssue } from "./gh"

// --- schemas ---

const BoardProject = Schema.Struct({
  id: Schema.String,
  name: Schema.String,
  key: Schema.String,
  counter: Schema.Number,
  github_repo: Schema.optional(Schema.String),
})
const BoardProjectList = Schema.Array(BoardProject)

const BoardIssue = Schema.Struct({
  id: Schema.String,
  project_id: Schema.String,
  title: Schema.String,
  description: Schema.optional(Schema.String),
  status: Schema.String,
  priority: Schema.String,
  assignee_id: Schema.optional(Schema.String),
  assignee_name: Schema.optional(Schema.String),
  labels: Schema.Array(Schema.String),
  created_at: Schema.String,
  updated_at: Schema.String,
  created_by_id: Schema.optional(Schema.String),
  created_by_name: Schema.optional(Schema.String),
  created_by_type: Schema.optional(Schema.String),
  github_issue_number: Schema.optional(Schema.Number),
  github_url: Schema.optional(Schema.String),
})
const BoardIssueList = Schema.Array(BoardIssue)

const BoardComment = Schema.Struct({
  id: Schema.String,
  issue_id: Schema.String,
  author_id: Schema.String,
  author_name: Schema.String,
  author_type: Schema.String,
  body: Schema.String,
  created_at: Schema.String,
})
const BoardCommentList = Schema.Array(BoardComment)

const BoardEvent = Schema.Struct({
  id: Schema.String,
  issue_id: Schema.String,
  event_type: Schema.String,
  actor_name: Schema.String,
  actor_type: Schema.String,
  timestamp: Schema.String,
})
const BoardEventList = Schema.Array(BoardEvent)

const VoidResponse = Schema.Struct({})

const ImportResult = Schema.Struct({
  imported: Schema.Number,
  skipped: Schema.Number,
  total: Schema.Number,
})

const ExportResult = Schema.Struct({
  exported: Schema.Number,
  files: Schema.Array(Schema.String),
})

// --- shared options ---

const issueId = Args.text({ name: "id" })

// --- project subcommands ---

const listProjectsCommand = Command.make("list", {}, () =>
  Effect.gen(function* () {
    const kernel = yield* KernelClient
    const projects = yield* kernel.get("/api/board/projects", BoardProjectList)

    if (projects.length === 0) {
      yield* Console.log("No projects found.")
      return
    }

    yield* Console.log(`${"Key".padEnd(10)} ${"Name".padEnd(30)} Issues`)
    yield* Console.log("-".repeat(50))
    for (const p of projects) {
      yield* Console.log(`${p.key.padEnd(10)} ${p.name.padEnd(30)} ${p.counter}`)
    }
  })
)

const projectName = Options.text("name").pipe(Options.withDescription("Project name"))
const projectKey = Options.text("key").pipe(Options.withDescription("Project key (e.g. PROJ)"))

const createProjectCommand = Command.make(
  "create",
  { name: projectName, key: projectKey },
  ({ name, key }) =>
    Effect.gen(function* () {
      const kernel = yield* KernelClient
      const project = yield* kernel.post(
        "/api/board/projects",
        { name, key },
        BoardProject
      )
      yield* Console.log(`Project created: ${project.key} — ${project.name}`)
    })
)

const projectsCommand = Command.make("projects").pipe(
  Command.withSubcommands([listProjectsCommand, createProjectCommand])
)

// --- issue subcommands ---

const projectId = Options.text("project").pipe(
  Options.optional,
  Options.withDescription("Filter by project ID")
)
const statusFilter = Options.text("status").pipe(
  Options.optional,
  Options.withDescription("Filter by status")
)
const assigneeFilter = Options.text("assignee").pipe(
  Options.optional,
  Options.withDescription("Filter by assignee ID")
)
const labelFilter = Options.text("label").pipe(
  Options.optional,
  Options.withDescription("Filter by label")
)
const limit = Options.integer("limit").pipe(Options.withDefault(50))

const listIssuesCommand = Command.make(
  "list",
  { project: projectId, status: statusFilter, assignee: assigneeFilter, label: labelFilter, limit },
  ({ project, status, assignee, label, limit }) =>
    Effect.gen(function* () {
      const kernel = yield* KernelClient
      const params = new URLSearchParams()
      params.set("limit", String(limit))
      if (Option.isSome(project)) params.set("project_id", project.value)
      if (Option.isSome(status)) params.set("status", status.value)
      if (Option.isSome(assignee)) params.set("assignee_id", assignee.value)
      if (Option.isSome(label)) params.set("label", label.value)

      const issues = yield* kernel.get(`/api/board/issues?${params.toString()}`, BoardIssueList)

      if (issues.length === 0) {
        yield* Console.log("No issues found.")
        return
      }

      yield* Console.log(`${"ID".padEnd(10)} ${"Status".padEnd(14)} ${"Priority".padEnd(10)} Title`)
      yield* Console.log("-".repeat(70))
      for (const i of issues) {
        yield* Console.log(
          `${i.id.slice(0, 8).padEnd(10)} ${i.status.padEnd(14)} ${i.priority.padEnd(10)} ${i.title.slice(0, 40)}`
        )
      }
    })
)

const issueTitle = Options.text("title").pipe(Options.withDescription("Issue title"))
const issueProject = Options.text("project").pipe(Options.withDescription("Project ID"))
const issueDesc = Options.text("description").pipe(
  Options.optional,
  Options.withDescription("Issue description")
)
const issuePriority = Options.text("priority").pipe(
  Options.withDefault("none"),
  Options.withDescription("Priority (none, low, medium, high, urgent)")
)

const createIssueCommand = Command.make(
  "create",
  { project: issueProject, title: issueTitle, description: issueDesc, priority: issuePriority },
  ({ project, title, description, priority }) =>
    Effect.gen(function* () {
      const kernel = yield* KernelClient
      const body: Record<string, unknown> = {
        project_id: project,
        title,
        priority,
        created_by_id: "shell",
        created_by_name: "gctrl-shell",
        created_by_type: "human",
      }
      if (Option.isSome(description)) body.description = description.value

      const issue = yield* kernel.post("/api/board/issues", body, BoardIssue)
      yield* Console.log(`Issue created: ${issue.id} — ${issue.title}`)
    })
)

const viewIssueCommand = Command.make(
  "view",
  { id: issueId },
  ({ id }) =>
    Effect.gen(function* () {
      const kernel = yield* KernelClient
      const i = yield* kernel.get(`/api/board/issues/${id}`, BoardIssue)

      yield* Console.log(`${i.id}: ${i.title}`)
      yield* Console.log("-".repeat(60))
      yield* Console.log(`Status:   ${i.status}`)
      yield* Console.log(`Priority: ${i.priority}`)
      yield* Console.log(`Assignee: ${i.assignee_name ?? "(unassigned)"}`)
      yield* Console.log(`Labels:   ${i.labels.join(", ") || "(none)"}`)
      if (i.description) yield* Console.log(`\n${i.description}`)
      yield* Console.log(`\nCreated: ${i.created_at}`)
      yield* Console.log(`Updated: ${i.updated_at}`)
    })
)

const moveStatus = Options.text("status").pipe(
  Options.withDescription("New status (backlog, todo, in_progress, in_review, done)")
)

const moveIssueCommand = Command.make(
  "move",
  { id: issueId, status: moveStatus },
  ({ id, status }) =>
    Effect.gen(function* () {
      const kernel = yield* KernelClient
      const issue = yield* kernel.post(
        `/api/board/issues/${id}/move`,
        { status, actor_id: "shell", actor_name: "gctrl-shell", actor_type: "human" },
        BoardIssue
      )
      yield* Console.log(`Issue ${issue.id} moved to ${issue.status}`)
    })
)

const assigneeId = Options.text("assignee-id").pipe(Options.withDescription("Assignee ID"))
const assigneeName = Options.text("assignee-name").pipe(Options.withDescription("Assignee name"))

const assignIssueCommand = Command.make(
  "assign",
  { id: issueId, assigneeId, assigneeName },
  ({ id, assigneeId, assigneeName }) =>
    Effect.gen(function* () {
      const kernel = yield* KernelClient
      const issue = yield* kernel.post(
        `/api/board/issues/${id}/assign`,
        { assignee_id: assigneeId, assignee_name: assigneeName, assignee_type: "human" },
        BoardIssue
      )
      yield* Console.log(`Issue ${issue.id} assigned to ${issue.assignee_name}`)
    })
)

const commentBody = Options.text("body").pipe(Options.withDescription("Comment text"))

const commentIssueCommand = Command.make(
  "comment",
  { id: issueId, body: commentBody },
  ({ id, body }) =>
    Effect.gen(function* () {
      const kernel = yield* KernelClient
      const comment = yield* kernel.post(
        `/api/board/issues/${id}/comment`,
        { author_id: "shell", author_name: "gctrl-shell", author_type: "human", body },
        BoardComment
      )
      yield* Console.log(`Comment added: ${comment.id}`)
    })
)

const eventsCommand = Command.make(
  "events",
  { id: issueId },
  ({ id }) =>
    Effect.gen(function* () {
      const kernel = yield* KernelClient
      const events = yield* kernel.get(`/api/board/issues/${id}/events`, BoardEventList)

      if (events.length === 0) {
        yield* Console.log("No events.")
        return
      }

      yield* Console.log(`${"Timestamp".padEnd(22)} ${"Type".padEnd(18)} Actor`)
      yield* Console.log("-".repeat(55))
      for (const e of events) {
        yield* Console.log(`${e.timestamp.padEnd(22)} ${e.event_type.padEnd(18)} ${e.actor_name}`)
      }
    })
)

const commentsListCommand = Command.make(
  "comments",
  { id: issueId },
  ({ id }) =>
    Effect.gen(function* () {
      const kernel = yield* KernelClient
      const comments = yield* kernel.get(`/api/board/issues/${id}/comments`, BoardCommentList)

      if (comments.length === 0) {
        yield* Console.log("No comments.")
        return
      }

      for (const c of comments) {
        yield* Console.log(`--- ${c.author_name} (${c.created_at}) ---`)
        yield* Console.log(c.body)
        yield* Console.log("")
      }
    })
)

const sessionId = Options.text("session").pipe(Options.withDescription("Session ID to link"))

const linkSessionCommand = Command.make(
  "link",
  { id: issueId, session: sessionId },
  ({ id, session }) =>
    Effect.gen(function* () {
      const kernel = yield* KernelClient
      yield* kernel.post(
        `/api/board/issues/${id}/link-session`,
        { session_id: session },
        VoidResponse
      )
      yield* Console.log(`Session ${session} linked to issue ${id}`)
    })
)

const issuesParent = Command.make("issues").pipe(
  Command.withSubcommands([
    listIssuesCommand,
    createIssueCommand,
    viewIssueCommand,
    moveIssueCommand,
    assignIssueCommand,
    commentIssueCommand,
    eventsCommand,
    commentsListCommand,
    linkSessionCommand,
  ])
)

// --- import / export ---

const importPath = Args.text({ name: "path" }).pipe(
  Args.withDescription("Directory containing .md issue files")
)

const importCommand = Command.make(
  "import",
  { path: importPath },
  ({ path }) =>
    Effect.gen(function* () {
      const kernel = yield* KernelClient
      const result = yield* kernel.post(
        "/api/board/import",
        { path },
        ImportResult
      )
      yield* Console.log(
        `Imported ${result.imported}, skipped ${result.skipped} (${result.total} total)`
      )
    })
)

const exportPath = Args.text({ name: "path" }).pipe(
  Args.withDescription("Directory to write .md issue files")
)
const exportProjectId = Options.text("project").pipe(
  Options.optional,
  Options.withDescription("Export only issues from this project ID")
)

const exportCommand = Command.make(
  "export",
  { path: exportPath, projectId: exportProjectId },
  ({ path, projectId }) =>
    Effect.gen(function* () {
      const kernel = yield* KernelClient
      const body: Record<string, unknown> = { path }
      if (Option.isSome(projectId)) {
        body.project_id = projectId.value
      }
      const result = yield* kernel.post(
        "/api/board/export",
        body,
        ExportResult
      )
      yield* Console.log(`Exported ${result.exported} issues:`)
      for (const f of result.files) {
        yield* Console.log(`  ${f}`)
      }
    })
)

// --- link-github ---

const linkGhRepo = Options.text("repo").pipe(
  Options.withDescription("GitHub repo (owner/repo)")
)
const linkGhProjectId = Args.text({ name: "project-id" }).pipe(
  Args.withDescription("Board project ID")
)

const linkGithubCommand = Command.make(
  "link-github",
  { projectId: linkGhProjectId, repo: linkGhRepo },
  ({ projectId, repo }) =>
    Effect.gen(function* () {
      const kernel = yield* KernelClient
      const project = yield* kernel.post(
        `/api/board/projects/${projectId}/github`,
        { github_repo: repo },
        BoardProject
      )
      yield* Console.log(`Linked project ${project.key} → ${repo}`)
    })
)

// --- sync helpers ---

const GhIssueList = Schema.Array(GhIssue)

/** Resolve a project by key and verify it has a linked GitHub repo. */
const resolveLinkedProject = (projectKey: string) =>
  Effect.gen(function* () {
    const kernel = yield* KernelClient
    const projects = yield* kernel.get("/api/board/projects", BoardProjectList)
    const proj = projects.find((p) => p.key === projectKey)
    if (!proj) return yield* Effect.fail({ _tag: "NotFound" as const, message: `Project "${projectKey}" not found.` })
    if (!proj.github_repo) return yield* Effect.fail({ _tag: "NotLinked" as const, message: `Project "${projectKey}" has no linked GitHub repo. Use: gctrl board link-github <project-id> --repo owner/repo` })
    return { ...proj, github_repo: proj.github_repo }
  })

const syncProject = Options.text("project").pipe(
  Options.withDescription("Project key (e.g. GCTL)")
)

// --- sync subcommands ---

const syncPullCommand = Command.make(
  "pull",
  { project: syncProject },
  ({ project }) =>
    Effect.gen(function* () {
      const kernel = yield* KernelClient
      const proj = yield* resolveLinkedProject(project)

      const [ghIssues, boardIssues] = yield* Effect.all([
        kernel.get(`/api/github/issues?repo=${encodeURIComponent(proj.github_repo)}&limit=100`, GhIssueList),
        kernel.get(`/api/board/issues?project_id=${proj.id}`, BoardIssueList),
      ])

      const existingGhNumbers = new Set(
        boardIssues
          .filter((i) => i.github_issue_number !== undefined)
          .map((i) => i.github_issue_number)
      )
      const newFromGh = ghIssues.filter((gi) => !existingGhNumbers.has(gi.number))

      yield* Effect.forEach(newFromGh, (gi) =>
        Effect.gen(function* () {
          yield* kernel.post("/api/board/issues", {
            project_id: proj.id,
            title: gi.title,
            description: gi.body ?? undefined,
            labels: gi.labels,
            github_issue_number: gi.number,
            github_url: gi.url,
            created_by_id: gi.author,
            created_by_name: gi.author,
            created_by_type: "human",
          }, BoardIssue)
          yield* Console.log(`  ← #${gi.number} ${gi.title}`)
        }), { concurrency: 1 })

      yield* Console.log(`Sync pull: ${newFromGh.length} new issue(s) from ${proj.github_repo}`)
    }).pipe(
      Effect.catchTag("NotFound", (e) => Console.log(e.message)),
      Effect.catchTag("NotLinked", (e) => Console.log(e.message)),
    )
)

const syncPushCommand = Command.make(
  "push",
  { project: syncProject },
  ({ project }) =>
    Effect.gen(function* () {
      const kernel = yield* KernelClient
      const proj = yield* resolveLinkedProject(project)

      const boardIssues = yield* kernel.get(`/api/board/issues?project_id=${proj.id}`, BoardIssueList)
      const unsynced = boardIssues.filter(
        (i) => i.github_issue_number === undefined || i.github_issue_number === null
      )

      yield* Effect.forEach(unsynced, (bi) =>
        Effect.gen(function* () {
          const created = yield* kernel.post(
            `/api/github/issues?repo=${encodeURIComponent(proj.github_repo)}`,
            { title: bi.title, body: bi.description, labels: bi.labels },
            GhIssue
          )
          yield* Console.log(`  → #${created.number} ${bi.title}`)
        }), { concurrency: 1 })

      yield* Console.log(`Sync push: ${unsynced.length} issue(s) to ${proj.github_repo}`)
    }).pipe(
      Effect.catchTag("NotFound", (e) => Console.log(e.message)),
      Effect.catchTag("NotLinked", (e) => Console.log(e.message)),
    )
)

const syncStatusCommand = Command.make(
  "status",
  { project: syncProject },
  ({ project }) =>
    Effect.gen(function* () {
      const kernel = yield* KernelClient
      const projects = yield* kernel.get("/api/board/projects", BoardProjectList)
      const proj = projects.find((p) => p.key === project)
      if (!proj) {
        yield* Console.log(`Project "${project}" not found.`)
        return
      }

      yield* Console.log(`Project: ${proj.key} — ${proj.name}`)
      yield* Console.log(`GitHub:  ${proj.github_repo ?? "(not linked)"}`)

      if (proj.github_repo) {
        const boardIssues = yield* kernel.get(`/api/board/issues?project_id=${proj.id}`, BoardIssueList)
        const synced = boardIssues.filter((i) => i.github_issue_number !== undefined)
        const unsynced = boardIssues.filter((i) => i.github_issue_number === undefined || i.github_issue_number === null)
        yield* Console.log(`Synced:   ${synced.length} issues`)
        yield* Console.log(`Unsynced: ${unsynced.length} board-only issues`)
      }
    })
)

const syncCommand = Command.make("sync").pipe(
  Command.withSubcommands([syncPullCommand, syncPushCommand, syncStatusCommand])
)

// --- board (parent) ---

export const boardCommand = Command.make("board").pipe(
  Command.withSubcommands([projectsCommand, issuesParent, importCommand, exportCommand, linkGithubCommand, syncCommand])
)
